cumulus_etl/etl/tasks/base.py

```diff
@@ -173,41 +178,60 @@ def _get_formatter(self, table_index: int) -> formats.Format:
 
         return self.formatters[table_index]
 
+    def _uniquify_batch(self, batch: list[dict]) -> list[dict]:
+        """
+        Drop duplicates inside the batch to guarantee to the formatter that the "id" column is unique.
+
+        This does not fix uniqueness across batches, but formatters that care about that can control for it.
+
+        For context:
+        - We have seen duplicates inside and across files generated by Cerner bulk exports. So this is a real
+          concern found in the wild, and we can't just trust input data to be "clean."
+        - The deltalake backend in particular would prefer the ID to be at least unique inside a batch, so that
+          it can more easily handle merge logic. Duplicate IDs across batches will be naturally overwritten as
+          new batches are merged in.
+        - Other backends like ndjson can currently just live with duplicates across batches, that's fine.
+        """
+        id_set = set()
+
+        def is_unique(row):
+            nonlocal id_set
+            if row["id"] in id_set:
+                return False
+            id_set.add(row["id"])
+            return True
+
+        return [row for row in batch if is_unique(row)]
```

Comment:
this feels like a good memory profiling target to me.

Reply:
This costs one extra batch size plus a set of of IDs, and is temporary. It's definitely adding to the problem, but this is (in my mind) an acceptable memory cost since it is reaped back so quickly. That is, my ideal ETL would peak at roughly 2x the batch size in memory.

Whether it's the most efficient, time or memory wise, is certainly an open topic.